### ISSUE TITLE
Server Save without Exit

### ISSUE THREAD

**Issue by friznit** [ALiVEOS/ALiVE.OS#197]
As an admin I want to save my multiplayer game to the ALiVE database without exiting the mission so that I can be reassured that the data has saved correctly and I have the opportunity to retry if it fails.

Given that I have a working ALiVE mission with database module and my persistent database details are setup correctly, when I initiate an Admin Server Save then the mission is saved to database.

Constraints: I assume the current Save & (forced) Exit is due to the need to save player profiles.  If so, a Save (& not exit) might have to exclude player profiles.  Note that players already have the option to save manually without exiting.

**Comment by Vegas** [comment]
This is an outstanding idea! I actually mentioned this exact request in the discord some days ago because I want/need the same thing.

**Having a forced exit is not ideal for a few reasons:**
1) Sometimes data fails to save (or has minor errors) and having a forced quit mission prevents a reattempt to save.
2) Sometimes when a player is crashing or lagging out, this affects the ability for ALiVE to save persistence. This has been a bug for a while.
3) Having this option will help with servers that run for extended periods of time because an admin can simply save when they feel they need to without returning all players to the lobby.

I have a few ideas/suggestions for this (as the 21st uses this feature every night):

- Perhaps it needs it's own button that is either accessible through the ALiVE menu (app data) or on Esc menu
- Save player profiles on exit (as Friznit mentioned) might also be handy to have as it's own button. So we have better control over what is saving and when.
- When actually using the "Server Save and Exit" function which triggers the ALiVE tablet to come up and start showing you save results, it would be nice to have it pause for a short time once completed before the mission begins the exit and the "da da duh daaaa" outro fade. The reason for a pause being handy is because when the save results flash by too fast, it is sometimes impossible to see what has result=OK and what failed or has an issue.
- As a side request, (maybe this is a bigger overhaul), but give mission designers greater control over what is saved. For example, it would be nice in the module settings if we could enable/disable the saving of objects inside containers/"inventory" of other objects. Might also be nice to have some sort of "flush" function where a server admin can reset the object persistence with a simple button/function and clean up messes/terrible bases that players have created without having to jump back to Eden.

Appreciate your time!

**Comment by friznit** [comment]
> This is an outstanding idea!

It was yours ;)  I read the convo on Discord and prompted you to log a ticket.

**Comment by coachwn** [comment]
Nice thought process and write up there.

**Comment by friznit** [comment]
@highhead is this possible with PNS saving?

**Comment by tupolov** [comment]
so the issue with saving to the DB was that the system was not designed for this use case. Reason being:
1. Original requirement was that we decided we would only save at end of mission to avoid unnecessary load and engine locks sending data and waiting for response
2. Bulk saving, we opted to save bulk data and not do individual record updates, this keeps the system simple and avoids data consistency issues (see Marker problems for why we avoided it!)
3. CouchDB index and record system uses a system of record hashes. We load the hash in at the start, it gets updated when we save, but the system isn't designed to cope with multiple record hash changes. I.e. we don't read back the new record hash after a save.
4. PNS wasn't possible when we coded the data system (edited)

so basically unless someone wants to recode the couchdb save, it won't be possible

ofc it is likely possible with PNS
